### ISSUE TITLE
Estimated schedule of completion

### ISSUE THREAD

**Issue by StuartIanNaylor** [linto-ai/linto-desktoptools-hmg#6]
Hi guys apols to bother but just wondering what your schedule is with the HMG as its a great tool but wondering when and what it will be.

What is due to v1.0 which notice is have a rebuild from the ground up and wondering what net architecture you may use or are you sticking with a basic GRU.
Also same MFCC as is the wait due to the likes of tensorflow.keras.layers.experimental.preprocessing / tensorflow.io and how that might pan out?
As that likely means when it out of your hands, but just wondering as been excited and waiting for some time now.

**Comment by Lokhozt** [comment]
Hello Stuart,

> Hi guys apols to bother but just wondering what your schedule is with the HMG as its a great tool but wondering when and what it will be.

1.0 "should" be ready with the core functionalities next week. Including dataset management, feature profiles, model profile, training, testing and export.

Release binaries (including ubuntu 20) might come alongside depending on my ability to reduce the executable size to an acceptable level.

> What is due to v1.0 which notice is have a rebuild from the ground up and wondering what net architecture you may use or are you sticking with a basic GRU.

As for the net architecture, release will keep the GRU based approach but will allow to add others later (I'm thinking CNN, attention, ...).

> Also same MFCC as is the wait due to the likes of tensorflow.keras.layers.experimental.preprocessing / tensorflow.io and how that might pan out?

No idea.

**Comment by StuartIanNaylor** [comment]
Cheers :) good news.

Its pretty easy to build anyway biggest problem with older machines is tensorflow and autokeras as spent some time finding which version would work.

GRU is fine but CRNN supposedly more accurate for far less ops. Less ops than a CNN.

Going on what Arm say...
https://github.com/ARM-software/ML-KWS-for-MCU

Tensorflow.io seem to be a lib for Mfcc and file operations.
But you can do like my hacky attempts with 2.4 https://github.com/StuartIanNaylor/simple_audio_tensorflow/blob/main/simple_audio_mfcc_frame_length256_frame_step128.py
You can drop sonopy...

**Comment by Lokhozt** [comment]
> Cheers :) good news.
> 
> Its pretty easy to build anyway biggest problem with older machines is tensorflow and autokeras as spent some time finding which version would work.
> 
> GRU is fine but CRNN supposedly more accurate for far less ops. Less ops than a CNN.
> 
> Going on what Arm say...
> https://github.com/ARM-software/ML-KWS-for-MCU

I took a look some time ago, without going further, i will investigate it more thoughtfully. Model scale is not the same though: 5k (for simple GRU) parameters to 229k (for CRNN). (https://arxiv.org/pdf/1703.05390.pdf). It's not specially a problem at the end.

> Tensorflow.io seem to be a lib for Mfcc and file operations.
> But you can do like my hacky attempts with 2.4 https://github.com/StuartIanNaylor/simple_audio_tensorflow/blob/main/simple_audio_mfcc_frame_length256_frame_step128.py
> You can drop sonopy...

How is it performance-wise ? I did some tests back a year or so and i remember that it was far behind numpy/scipy based implementations. As for sonopy, it will not be in the 1.0 as it lacks mfcc steps and has implementation errors.

**Comment by StuartIanNaylor** [comment]
Yeah I did the sonopy performance tests and then changed some of the parameters and reran. Something is not right about sonopy and I feel its currently only highly performant because it may be skipping bins or something as if you do change from the default parameters the speed increase over other libs is huge and dubious.
Not sure what or why?

I am really interested in distributed wide array microphones where KWS broadcast till silence is really important to lower network traffic.
All beamformers suffer badly when passing through a predominant noise field and rather have a single microphone packed wit DSP simple distributed microphones can be more effective as near voice / far noise can be assured by placement of a distributed array.
To make things cost effective its low end ESP32/microcontrollers to things like a PiZero/RK3308 so what was said about latency and ops with a CRNN sounded like it was beneficial.

Apols about my coding but with tensorflow 2.4 I just added the MFCC conversion to the Spectrogram function and its so performant I am wondering if the code is wrong.
Since 20 years of MS I don't code anymore but after waiting for a while being new year I had decided that maybe I should explore something else so hence hearing HMG will be released soon is a relief to my ability of Python programming.

I think you will have to take a look as compared to what you had with the last working version of HMG the above is much faster.

**Comment by StuartIanNaylor** [comment]
PS I did the same again with pruning the google command set and wondering if this would be a addition to HMG as a sperate window?

What I am doing is loading all directories as there own label and then running inference on each label wav against that label and pruning all that have low softmax score (I did < .3)
Its a long winded process but at least much easier than doing it manually via failures.

By doing it even the poor CNN I have been playing with gained about 5% accuracy without too much sample loss.
My sample count is down to 7700 but I did hit 92% a couple of times.
